### ISSUE TITLE
Support Facebook's GraphQL

### ISSUE THREAD

**Issue by pilwon** [rethinkdb/rethinkdb#3711]
Facebook announced **GraphQL** at the [React Conf '15](http://conf.reactjs.com/) and the spec will be released sometime in the near future.

* Presentation: https://www.youtube.com/watch?v=9sc8Pyc51uU
* FAQ: https://gist.github.com/wincent/598fa75e22bdfa44cf47

It could be a powerful add-on to RethinkDB (either on the client driver or server) providing an alternative, efficient way to construct complex, nested queries from multiple tables in a single query. It could also potentially bring a lot of React users, especially when it is combined with the power of recently announced real-time changefeed feature.

**Comment by danielmewes** [comment]
Hi @pilwon, that's an interesting suggestion.

I think we're going to concentrate on ReQL for the time being (and possibly add graph features, see https://github.com/rethinkdb/rethinkdb/issues/3142), but embedding another query language will eventually become an interesting option.

I wonder if we could add RethinkDB support to React independently of supporting GraphQL. That would be one for @deontologician , I haven't looked into it at all yet.
Another option might be a client-side "wrapper" library that translates GraphQL queries into ReQL.

**Comment by Agamennon** [comment]
GraphQL would be enormous boost to rethinkdb adoption, if it could be done close to release it will be the best db option for anyone using Facebook stack, and that mind-share is big and going fast

**Comment by thejsj** [comment]
@pilwon, @Agamennon Isn't what you really want an adapter for GraphQL that sits on top of RethinkDB? Would this be a project be outside the scope of the database? You can't, for example, query RethinkDB from the client (and you probably don't want to either) and GraphQL would always be queried from the client.

**Comment by pilwon** [comment]
@thejsj An adapter transforming GraphQL into ReQL statement is definitely one way, but more preferred approach I hope to see is a new `r.graphql(<raw-graphql-statement>)` that can receive raw GraphQL statement and internally process and return back most optimized results so the web server can simply forward them to the client. Added benefit with this approach is you only implement GraphQL once in the database engine itself and all RethinkDB client libraries will benefit from any subsequent updates without code updates. (no need to create and update adapter for every language RethinkDB client library supports)

As @Agamennon also mentioned above, this single feature implementation can potentially bring massive number of React developers and instantly turn RethinkDB into a de-facto standard database for React applications. Imagine what positive reputation it will add when it also integrates seemless with changefeed too.

This single endpoint approach really seems taking off right now. There's also a promising project called [falcon](https://www.npmjs.com/package/falcor) by Netflix trying to solve a similar problem. I think jumping into this early will help RethinkDB stay ahead of the game and win over its competitors easily.

**Comment by jbroadway** [comment]
Facebook just posted their [GraphQL reference implementation](https://github.com/graphql/graphql-js). It would be really cool to see a GraphQL to RethinkDB layer, or a RethinkDB mixin for React like Firebase has with [ReactFire](https://github.com/firebase/reactfire).

**Comment by marshall007** [comment]
After skimming through the [draft specification](https://github.com/facebook/graphql) it appears as though GraphQL relies pretty heavily upon schema definitions through its own type system in order to validate queries and to support reflection. As a result, I don't think RQL itself could support a term like `r.graphql(<statement>)` because the statement needs to be interpreted in the context of application-specific types.

However, as @danielmewes mentioned, it would probably be possible to implement this as a wrapper around RQL (similar to [`thinky`](https://github.com/neumino/thinky)) which could be made aware of user-defined schemas.

**Comment by divmain** [comment]
I saw this added the other day, might be similar to what you're looking for @jbroadway.
https://github.com/mikemintz/react-rethinkdb

**Comment by jbroadway** [comment]
Thanks @divmain! Just saw that pop up on Hacker News this morning.

**Comment by Agamennon** [comment]
So this is what Facebook is selling, you have a relay store that talks to a Graphql server, relay assembles the query to the Graphql server that can answer that request accordingly, now why do i believe rethink could profit from this? my answer is two fold:

1: Graphql on its own, is a much better way to query data from the server then thinking about the exact structure of how the data is stored on the server, in practice you never use tables or other "sql like" abstractions, what you want is just a specific Json, so Graphql on its own, even when not coupled with Ralay or other React fluff, stands on its own.

2: By having an internal implementation of Graphql, Rethinkdb could optimize and do other stuff that drivers would need to do, and do that inside the server without back and forth information being exchanged with a potential rethink-graphql server driver, and they can only do that by inferring on the structure of the user data, and requiring the user to probably provide some mappings (this last part might be inevitable)

So, IMHO, Graphql should be taken seriously! rethinkdb would profit tremendously, as querying data from it would be intuitive to web developers, the old ways of writing backed code is over, rest or rpc specific code, is probably going to be obsolete, lets adjust accordingly.

**Comment by geddski** [comment]
GraphQL does sound promising. I think it would be premature though to bake in support for it (if that were even possible) into RethinkDB. A solid adapter would probably be awesome enough.

**Comment by contra** [comment]
@geddski Got any pointers on where somebody implementing an adapter should start?

**Comment by andrerpena** [comment]
Facebook has just released Relay: https://github.com/facebook/relay

**Comment by geddski** [comment]
@contra I'm looking into the same question currently.

**Comment by rattrayalex** [comment]
As basically "a random member of the React community", I'm seeing a lot of shops switching (actively, soon, or planned) from Angular/whatever to React. I myself recently did a big switch from Angular to React and the results are just great. Going straight to Relay as well at the same time would be a massive win, the problem is that there aren't any GraphQL implementations out yet. 

RethinkDB seems poised to be a top candidate to support it faster and better than any others, since it's already so much closer to what the client is asking for than, say, PGSQL (which would otherwise be my first choice). 

While a community adapter is a totally sane approach to this, and in fact I nebulously hope to experiment with building one myself, native official support from the RethinkDB team could really push things over the edge for a lot of teams who are starting out or rewriting and want to be building with best-of-breed technologies.

tl;dr, I agree with @Agamennon that official support for GraphQL could lead to a big growth boost for RDB, though a community adapter would be "fine" too.

**Comment by coffeemug** [comment]
Cofounder @ Rethink here. Firstly, a disclaimer: I haven't programmed in React beyond doing a very simple toy project, so there's a possibility I'm going to say something stupid.

From a technical perspective I think there are a few ways of doing this:

1. An `r.graphql` command. GraphQL does seem to depend on the schema to evaluate queries, but I don't think it's a problem. The command could just take a schema, and we could cache it in the db so to avoid the performance hit of the user sending it every time.
  - Pros: one implementation to rule all them languages; efficient
  - Cons: server team is currently a bit resource constrained (tons of users/customers want very different things); users still need to build a backend; baking a very new tech into db is a bit risky
2. Rethink's HTTP server could support GraphQL queries directly (potentially built on top of `r.graphql`)
  - Pros: you can build react apps without building a backend at all; this would be frakkin' awesome
  - Cons: we'd have to solve security, which is hard
3. Ship it as an adapter on top of RethinkDB via a language library
  - Pros: probably quite easy to write a GraphQL->ReQL adapter
  - Cons: We'd start in JS, but what about all the other languages?
4. Ship it as an adapter on top of RethinkDB via a proxy project (possibly on top of a language library)
  - Pros: also easy to do; you don't need a backend; security is easier
  - Cons: not as nice a user experience; worse from the marketing POV

I think this is super-important and we've got some serious decisions to make. The team is going rafting tomorrow to celebrate the Raft release :grin: but I'll see if we can carve out some time and discuss this ASAP.

**Comment by pilwon** [comment]
@coffeemug It's great that you see this as a tremendous growth opportunity as well! :+1:

For user's perspective, `Option 2` is no doubt the most compelling option, however I feel `Option 1` is (almost) as good as `Option 2`. You just need to show React developers how easy and fun it is to play with `r.graphql` in the data explorer. Building a backend is not a major burden if you also provide a rdb wrapper for [express-graphql](https://github.com/graphql/express-graphql) (officially supported by FB).

In my opinion, officially shipping an adapter (`Option 3` or `Option 4`) may not bring the result you'd expect. There will most likely be tons of other open source projects like that for all sorts of databases so it will be hard to stand out even with a first-mover advantage. It might be better to just leave it up to the community.

**Comment by pilwon** [comment]
On their [blog post](http://facebook.github.io/react/blog/2015/08/11/relay-technical-preview.html):
> Real-time updates. In collaboration with the GraphQL community, we're working to define a specification for subscriptions and provide support for them in Relay.

The Relay team is already working on the spec to add real-time updates functionality. This clearly seems like an opportunity for RethinkDB as there aren't many databases out there that can even dream of supporting it natively... :feelsgood:

**Comment by rattrayalex** [comment]
Sounds good to me @coffeemug and +1 to @pilwon 's comments. Have a great time rafting!! You guys deserve it!

**Comment by contra** [comment]
Option 1 or Option 3 are best IMO 🌴

**Comment by andrerpena** [comment]
@coffeemug and others, I'd like to say that the attitude I'm seeing here in this thread is one of the reasons why I'm building my product on top of RethinkDB even though it's not a market leader and even though it does not have support for Windows yet. You guys really care for innovation and promptly adopting new technologies that look promising, for the sake of progress. This is what technology is all about. Thanks everyone for your energy. RethinkDB rocks because of you.

**Comment by geddski** [comment]
I like the no backend options even though they would be harder to implement. Fits with the Relay mentality of focusing on building your app, not writing services. 

I think any effort here is awesome. Just one more reason to love the RethinkDB team. Just don't get acquired by FB ;)

**Comment by rattrayalex** [comment]
Possibly relevant, the RisingStack claims to be working on this with their Graffiti project ( https://github.com/RisingStack/graffiti ) with an adapter in the works for the Thinky rethinkdb orm.

**Comment by ide** [comment]
I'd like to echo the sentiment that RethinkDB could be a really good fit for Relay in particular because of your support for realtime queries.

@coffeemug - my inclination is that Option 3 (library) leading towards Option 1 (engine support) is a good path forward. There are many GraphQL queries that don't map to ReQL in an obvious way (ex: "get me this person's posts... that I'm allowed to see" -- the privacy constraint is implicit and GraphQL has the notion of the current user but ReQL doesn't), which is why I believe it's so important to have a library that lets us implement parts of a GraphQL query with our own code that calls into RethinkDB. Native support for GraphQL is a nice optimization underneath that. I would strongly advise against Option 2 because of the security issue you brought up... you really want complex security and privacy logic to live in the application layer.

**Comment by coffeemug** [comment]
@ide -- cool, thanks for the feedback. Do you have an example of how to do security with GraphQL you could point me to? I couldn't find anything online (other than a vague "do it yourself").

Wouldn't the developer have to analyze every GraphQL query themselves to do security? That sounds like a huge burden, and it feels strange that security provisions aren't in the spec. I feel like I'm missing something, but I can't seem to find any information on this.

**Comment by compedit** [comment]
This *might* be helpful, it's from an [interview with some of the relay team members](https://devchat.tv/js-jabber/152-jsj-graphql-and-relay-with-nick-schrock-and-joe-savona-)

> CHUCK: ...as far as security goes, I don’t want just anybody formulating GraphQL queries to send to my system. Is the structure inherently more secure? Or do you just secure your endpoint the same way you it with anything else? So, OAuth, authentication headers, HTTPS.

> NICK: Yeah, we just use existing authentication.

> NICK: Basically we effectively grab an access token from out external platform and we ensure that’s from a Facebook app. And then we piece out and then we just execute the string. So, GraphQL’s an explicitly higher level protocol than HTTP. It is an application layer protocol. That’s one of the things I find actually strange about REST, is the coupling of HTTP and the application level protocol. So, the authentication mechanism is completely orthogonal to GraphQL.

**Comment by danielmewes** [comment]
Thanks @compedit.
So that covers authentication, though it doesn't really cover authorization, i.e. assigning which users have access to which parts of the data. It seems to me that there's no standard for this yet as far as GraphQL is concerned?

**Comment by ide** [comment]
@coffeemug I'm not aware of an example project that demonstrates security and privacy but hopefully I can shed some more light on how it's done with GraphQL.

The authorization is implicit and not part of GraphQL. When a client (e.g. browser, mobile device) sends a GraphQL query to the server, it also sends its auth cookie so that the server knows who's making the request. Nothing new here. It's your job when constructing the GraphQL response to make sure the server doesn't send back information that the client isn't allowed to see.

One way to do this is to have a DAO library that is aware of the notion of a "viewer". Whenever you implement a new DAO data type (e.g. User, Photo, Comment, Place) it's your job to write logic that determines whether the viewer is allowed to access an object of that data type. This also applies to the connections between objects; a viewer may be allowed to see a User and a Place but might not be allowed to see that the User has a check-in at that place. This logic differs from app to app and it's quite hard for a framework or service to anticipate all the use cases.

With these DAO data types correctly implemented, you can safely write code like `User.get(viewer, userId).getFriends().map(friend => friend.getHometown())` and know the result honors the privacy of the specified user and their friends with respect to the viewer.

Finally, a server-side GraphQL implementation fulfills the query mostly by calling into the DAO library. I say "mostly" because some parts of the query get their data elsewhere (e.g. localized strings).

Where I think a RethinkDB+GraphQL library helps out is: good support for converting RethinkDB results to GraphQL data types, support for the reverse for handling GraphQL mutations (aside: similar to how the DAO layer had privacy logic for reads, it needs permissions logic for writes), and a story around how to use RethinkDB's changefeeds to build GraphQL subscriptions.

**Comment by danielmewes** [comment]
Moving to subsequent because this will not be ready for 2.2. However we are going to put up a concrete proposal in the next days and will start working on this soon to get it into 2.3.

**Comment by kevinSuttle** [comment]
I'm _very_ interested in this being added. Definitely keep everyone in the loop, please.

**Comment by mike-marcacci** [comment]
As a reasonably active member of the RethinkDB community and a very new user of GraphQL, this issue got me very excited at first. However, I have developed some very real reservations about baking this into the RethinkDB core, and I'm a bit concerned that this initiative misinterprets the purpose of GraphQL.

In the past, I have rolled my own model systems on top of RethinkDB with similar schema and joining abilities as GraphQL. Despite the amazing flexibility of RQL and RethinkDB, there are implementation complexities that are not solved in RethinkDB and are out of the scope of GraphQL itself. One of these problems has already been discussed a bit in this thread – authorization – but the others have not: schema enforcement (writes, migrations), data structure (embedded collections vs. new tables, relationship structure), 

Issues
------

### Authorization
Authorization is an especially complex beast because each application is going to have completely different rules and structures. It almost goes without saying that this belongs in the application layer, and is for the most part outside the scope of this issue. However, it's important to keep in mind that this flies in the face of any advantage gained by native RethinkDB support, as the application needs to interpret the GraphQL request anyhow (and might as well generate RQL).


### Data Structure: Relationships
One of the key features of GraphQL is its ability to traverse relationships – it is called **Graph**QL after all – and at the moment RethinkDB doesn't have a way to internally mark relationships between documents. 

A solution for this might be to add an "edge" index type that links two tables (which sounds really cool as I think about it... perhaps I'll think it through more and make an issue) or a new `$reql_type$`. But without such a core feature, GraphQL is not possible.


### Data Structure: Collections
A key feature of document databases is the ability to store large, complex objects (or collections of objects) embedded inside documents. A graph database, on the other hand, emphesizes separation of objects and relating them.

Rethinkdb is a document database in true form, and on several occations I have seen performance improvements by embedding otherwise singlarly-relational data inside one document. More importantly, without [atomic multi-document updates](https://github.com/rethinkdb/rethinkdb/issues/4598) this is often a necessary design decision for data consistency.


### Schema Enforcement: Writes
Obviously, a schema system (which I am TOTALLY for as a feature in its own rite) needs to be implemented for GraphQL types to work. This is a totally separate discussion: should this be GraphQL or json-schema? How should these be configured? What happens when a write fails checks? What happens to invalid documents when the schema is added? How does one change a schema?


### Schema Enforcement: Migrations
In line with the last comment (but I think it deserves its own heading) is the issue of data migrations on schema changes. Is it safe to implicitly coerce data on its way out? Should we lock tables and perform alters à la typical SQL? Should we continue to serve old format, but inforce new writes?

These are all HUGE decisions that are prerequisite for GraphQL to become a core part of the database. While I am in favor of multi-document transactions, graph-database style edges and schema enforcments as individual features, they aren't (as far as I know) event close to being implemented.


The Purpose of GraphQL
----------------------
Please forgive me if I'm wrong about this (I certainly don't work at Facebook) but my understanding is that GraphQL was designed to communicate with an API that contains application logic. While it could certainly be used for querying a logicless API (in this case a RethinkDB database), it feels odd to embed it in a database's core.

This is how I imagine most moderately complex GraphQL-based apps will end up working:

```
     _____________
    |  RethinkDB  |
     ‾‾‾‾‾‾|‾‾‾‾‾‾
           |------------------------ RQL
 __________|__________
|   GraphQL Parsing   |
|         and         |
|  Application Logic  |
 ‾‾‾‾‾‾‾‾‾‾|‾‾‾‾‾‾‾‾‾‾    SERVER
-----------|------------------------ GraphQL
     ______|______        CLIENT
    |   Browser   |
     ‾‾‾‾‾‾‾‾‾‾‾‾‾
```

To even begin adding logic, the application needs to interpret GraphQL and act accordingly, and at this point it might as well build a RQL query.


Final Thoughts
--------------
I really hate to be such a downer, but the more I've thought about this, the more the idea concerns me.

A quick comparison: When I tried out [Meteor](https://www.meteor.com/) for the first time, I was amazed by how easy it is to write the ubiquitous To-Do app: I fell in love with the ability to perform database queries from the front-end. But once I tried to secure my application, I was forced to rewrite all my logic as back-end functions, and use the front-end to call those... and before I knew it, I had no more front-end queries. I had lost everything that drew me to meteor in the first place. All their hard work on minimongo was left unused and wasted.

I feel like this feature, while really awesome for ultra-simple To-Do style apps, has very little value for even moderately complex applications. Instead of coupling this to the RethinkDB core, I think it's best left to the application layer, or as a database plugin (which is a REALLY cool idea, btw).

The RethinkDb team could instead focus on implementing general-purpose features like transactions, schemas and document edges/relationships that would make writing a GraphQL plugin/library easier and more performant.

**Comment by ide** [comment]
@mike-marcacci great writeup. Agreed that we want to have a high ceiling that supports building real applications beyond demos. There are a handful of very (overly?) client-centric frameworks/services that make it easy to get started but hard to evolve as you mention. In that vein, automatically exposing RethinkDB documents via GraphQL could demo very well and be great for marketing but not be what you actually want for professional work. GraphQL is generally not intended to have a direct connection to RethinkDB or MySQL or any other traditional database (that said, it would be sweet for a DB admin tool where you do want a direct view of the database).

Where I believe RethinkDB can help is providing core features and helper libraries that make it easier for developers to build a GraphQL layer on top of it. I totally agree with your sentiment when you wrote, "The RethinkDb team could instead focus on implementing general-purpose features like transactions, schemas and document edges/relationships that would make writing a GraphQL plugin/library easier and more performant." We'll have to build apps with RethinkDB and GraphQL to learn what those general-purpose features are.

Also I'm still excited by the prospect of using RethinkDB's changefeeds to power live GraphQL subscriptions =)

**Comment by mike-marcacci** [comment]
@ide thanks for the quick response! I completely agree with your perspective on this, especially with regards to using changefeeds as a backing for GraphQL subscriptions. In my opinion, RethinkDB is one of the best tools available for this kind of project, in a major way due to RQL's expressiveness and composibility.

If anyone wants to start work on a GraphQL-to-RQL library, I'd be more than happy to contribute when I can!

**Comment by danielmewes** [comment]
Very interesting writeup @mike-marcacci, thanks for sharing your thoughts.

I think there is still a big demand for writing web applications without needing to write any backend code. Firebase for example is built almost entirely on this assumption, and you can do a lot of cool things with it. You're certainly right that there are limits to this approach, and eventually a complex application will probably require a special backend of some form. However there are also many applications that can work perfectly well at scale without an application-specific backend, assuming the tools provided by the database for this purpose are flexible and powerful enough.

You are right that GraphQL is different enough from ReQL, and it would add a lot of complexity to support it as equal citizens next to each other directly in the server. We're a couple of steps away from supporting full authorization, references and schema enforcement directly in the server, and as you point out all of these things are required for properly supporting GraphQL.

However I think there's a lot of potential in supporting GraphQL through our upcoming plugin infrastructure. In fact it's a prime example of what the plugin system would be good for.
Essentially all the GraphQL-specific logic would be implemented in JavaScript and run in the plugin. It would perform schema validation, follow relationships, and provide authorization. It would then translate each GraphQL query into one or multiple ReQL queries, and issue them to the RethinkDB server. We can utilize Facebook's reference demo implementation, and likely only need to add authorization and the ReQL translation layer.

This also has the nice benefit that GraphQL support doesn't put much of an additional burden on the core server development. We might end up having to implement a few additional features in ReQL to fully support all GraphQL features, but those would likely benefit a wide range of use cases and wouldn't actually be specific to GraphQL.

With us providing the GraphQL plugin first hand, we will be able to provide a particularly good user experience, in that we can package the underlying RethinkDB server and the GraphQL plugin up nicely. The goal is that you need to install and administer exactly one piece of software, and you can get started writing your applications without being concerned about the backend.
Internally, the RethinkDB server and the GraphQL plugin are really two separate pieces of software, but as a new user you don't need to be worry about that.
The other nice thing is that since the plugin will be very easy to modify and extend, more advanced users can start building their own extensions on top of it, and will get instant full access to the expressiveness of ReQL at that point.

I hope that makes sense.

We'll hopefully get started with basic work around a GraphQL plugin/layer in the next couple weeks. Any contributions would be amazing at that point! :-)

**Comment by leebyron** [comment]
Hello, I'm one of the primary authors of GraphQL.

The intuiting done so far on this thread is really good. I feel like I don't need to add much to this discussion.

@mike-marcacci's ASCII diagram above is a very accurate representation of how we view GraphQL and how GraphQL is used at Facebook: Client applications send requests to application servers using GraphQL, and those application servers are on their own to fulfill them, often using a combination of database queries (like ReQL), cache hits, and unhindered business logic (e.g. privacy/access rules or derived data).

However, I would like to toss into this discussion that Perfect should not be the enemy of Good here. I think there is huge value in @coffeemug's "Option 1" above.

I believe there is a continuum of needs and costs between "RethinkDB should fulfill GraphQL directly" to "RethinkDB should not know anything about GraphQL" along which lies a trade-off threshold. An ideal set of solutions pick a few points along this threshold and should be easy to move between as your needs evolve:

When people first get started building something new, the only criteria that matters is speed. Product criteria is often unknown because the product is unknown. Removing any overhead costs is critical to the ability to explore and prototype quickly. It's at this point where RethinkDB serving GraphQL directly is extremely high value. This will allow you to immediately take advantage of the client-side developer efficiency wins that GraphQL promises while avoiding the upfront cost of designing a schema.

Of course, directly serving GraphQL has limitations. As already mentioned in this issue: privacy/access rules are not something RethinkDB does directly, nor could it as many products (incl Facebook) require arbitrary logic and sometimes the fetching of additional information to determine these visibility rules. There's also no way (nor should there be) for RethinkDB to "fall through" to other systems in the case where your system evolves to the point that multiple backend services are necessary. None of these restrict the immediate value of being able to use GraphQL when your application is still simple.

The critical difficult issue to solve is how to build GraphQL integration into RethinkDB in such a way that as your product needs evolve and become more complication, you can iteratively move the responsibility for fulfilling GraphQL from RethinkDB into your application server logic - effectively leaving your client applications sending GraphQL unchanged, but opening these possibilities.

**Comment by coffeemug** [comment]
Hi @leebyron (I'm Slava btw, the red-haired guy that had lunch with ya'll a few weeks back). Thanks for chiming in!

Roughly, our current plan is as follows (though it may evolve as we get deeper into the implementation):

- Add an extension architecture to RethinkDB that implements a minimum possible featureset to make extensions useful.
- Build a GraphQL extension that will ship with RethinkDB out of the box.

The goals of the GraphQL extension are:

- Serve GraphQL directly from RethinkDB, so users can go from nothing to working React apps very very quickly, without writing any backend code.
- Implement a basic (but hopefully sufficiently flexible) security model so people can ship real apps into prod.
- Have a design that allows for a piecemeal migration to custom business/security logic for the relevant subsets of GraphQL queries as the users's apps get more complex. I'm not 100% certain how we'll do this yet, but it shouldn't be *too* hard.

I think this approach will turn out to be super-valuable to people and won't bloat RethinkDB. I'm *really* excited about it; can't wait to get it out the door!

**Comment by marshall007** [comment]
> There's also no way (nor should there be) for RethinkDB to "fall through" to other systems in the case where your system evolves to the point that multiple backend services are necessary.

@leebyron flipping this around, is GraphQL composable in any way? i.e. is it possible to configure a GraphQL server to read the schemas from and act as a proxy to other GraphQL endpoints?

**Comment by leebyron** [comment]
> @leebyron flipping this around, is GraphQL composable in any way? i.e. is it possible to configure a GraphQL server to read the schemas from and act as a proxy to other GraphQL endpoints?

GraphQL is as composable as your programming language allows! You could absolutely call other GraphQL endpoints from a GraphQL server. I don't think I've seen this actually done before, but there's no reason it couldn't be done.

**Comment by leebyron** [comment]
I should say, the important reason for why this is the case, and why I think it's not reasonable for RethinkDB to do this is that GraphQL is a way to write a mapping from field names to arbitrary functions in your programming language of choice. You can think of it as a sort of nested RPC. In contrast RethinkDB, like other DBs are not mappings to arbitrary logic but to specific storage engines. GraphQL was not designed to map directly against a storage engine in this way.

**Comment by ide** [comment]
> @leebyron: The critical difficult issue to solve is how to build GraphQL integration into RethinkDB in such a way that as your product needs evolve and become more complicated, you can iteratively move the responsibility for fulfilling GraphQL from RethinkDB into your application server logic without some massive migration effort

> @coffeemug Have a design that allows for a piecemeal migration to custom business/security logic for the relevant subsets of GraphQL queries as the users's apps get more complex. I'm not 100% certain how we'll do this yet, but it shouldn't be too hard.

I believe in these points strongly enough that I want to reiterate how important it is for RethinkDB+GraphQL to have a high ceiling that accommodates many needs. As a counterexample, Parse's model makes it difficult to build applications whose security & privacy logic doesn't neatly fit into an ACL-shaped hole. You end up needing to write Cloud Code at which point you're better off running your server-side code on your own servers in many ways. This sounds similar to @mike-marcacci's experience with Meteor if I understood him correctly.

I do love the idea of being able to simply drop some GraphQL fragments into my React components and have them load data from RethinkDB. That is awesome! I guess I have strong feelings that this isn't the goal but instead a path along the way to building applications that have custom logic between GraphQL and RethinkDB. It's a means to an end.

@marshall007 it's a cool idea. I think it'd be neat to write `me { facebookFriends { nodes { instagramPhotos { likes: { nodes: { twitterUsername } } } }}`. You get the idea :)

**Comment by mike-marcacci** [comment]
Hi guys, I'm completely in favor of this as an officially supported extension – both for the ability to immediately use GraphQL for prototyping and to showcase the RethinkDB extension system! It really feels like a sweet spot that makes RethinkDB uniquely qualified as both a prototyping and production solution. 

That said, I do have one remaining reservation, which I've been struggling to refine over the past couple days, but here goes:

To me, it's important that this kind of alternative interface doesn't take advantage of RethinkDB features that are not generally available through RQL or implementable in the application. For example, it would be odd for the GraphQL interface to support atomic bi-directional relationships without also providing a special type of index or supporting multi-document transactions over RQL.

The idea here is that there needs to be a clear, iterative path from prototype to production, where core guarantees and features aren't lost when your models are accessed through RQL half of the time and GraphQL the other half.

I think I'm more or less echoing the sentiment from everybody else here, but I just wanted to reiterate them here.

**Comment by mike-marcacci** [comment]
And it looks like @ide beat me to it while I was typing :) What he said exactly.

**Comment by danielmewes** [comment]
@mike-marcacci I agree 100%. ReQL is going to remain the primary query language. With our current plan for the plugin infrastructure, what you say is actually enforced. The GraphQL plugin simply has no other choice but to use ReQL queries to communicate with the database.
If it turns out that we need to add additional functionality to the core server itself, we'll make sure that that functionality is exposed through ReQL, and we'll try to make it as generic as possible (i.e. not limited to the specific use case that we need it for in GraphQL).

About extensibility and being able to replace the "default" endpoints in GraphQL as the demands of an application increase:
These are very valid points, and I hope we can provide something to easily "overwrite" a GraphQL endpoint.
What I mean by that is that by default, we're essentially going to map each GraphQL endpoint in the schema to a document or table in RethinkDB. That way you can store and retrieve data through GraphQL out of the box, without writing custom backend code. I think we could provide an API that allows you to inject custom callbacks for a certain endpoint. In the simplest case, you can just put another ReQL query there. In more complex cases, you can perform any type of additional authorization, or query out to other databases or backends.

One thing that's really nice about putting GraphQL support into a plugin, is that all of that logic is going to be plain JavaScript. The code of the plugin itself will be very easy to access and to modify. Even if we forget to provide a specific API for customization, you can simply make a copy of the default GraphQL plugin, and change the relevant pieces in the code.
None of this requires special knowledge about RethinkDB internals.

**Comment by pbowyer** [comment]
Not wishing to derail the thread, but if you're considering bundling GraphQL support with RethinkDB, why not Falcor (from Netflix) support instead or as well? Either would be good, but I'm aware Falcor isn't mentioned in this thread so wanted to ensure people are aware of it.

**Comment by marshall007** [comment]
@pbowyer as @coffeemug mentioned above, GraphQL support won't be baked in:

> Roughly, our current plan is as follows (though it may evolve as we get deeper into the implementation):
> 
> - Add an extension architecture to RethinkDB that implements a minimum possible featureset to make extensions useful.
> - Build a GraphQL extension that will ship with RethinkDB out of the box.

So while there will be an *official* GraphQL plugin, the extension system should allow you to write a Falcor (or any other third-party) integration in much the same way.

**Comment by coffeemug** [comment]
@pbowyer -- good question :smile: I actually spoke to the Netflix team that's building Falcor, and we do in fact have plans to build a Falcor plugin. I don't think there is an issue for it in the tracker, but the idea behind the plugin system is to allow support for a wide range of data protocols, and let the users pick whichever protocol makes the most sense for them.

**Comment by ansarizafar** [comment]
Amazing discussion. Is there any ETA on the release yet?

**Comment by coffeemug** [comment]
No ETA yet; but I think what we've got in store is going to be amazing. Can't wait to share more details, but we have to do a bit more technical work before we can impress the community.

**Comment by pbowyer** [comment]
@marshall007 @coffeemug Awesome, thank you!

**Comment by opsb** [comment]
First of all, this sounds like an exciting direction for rethinkdb. Something that I haven't  seen mentioned yet is how changefeeds will be exposed. I haven't seen this tackled anywhere with regards to GraphQL. Assuming updates were sent over a websocket, what might they look like? 

Seeing as the results of GraphQL queries are json structures perhaps JSON-Patch could be a good fit? (related to https://github.com/rethinkdb/rethinkdb/issues/3575)

The challenge I see though is that GraphQL results don't require ids to be included with embedded types so addressing the correct point in the graph would require json paths that included array indexes which doesn't seem robust enough.

**Comment by michaelmior** [comment]
I don't think GraphQL address streaming results. It seems reasonable that if this is something people are interested in and having IDs be included makes things a lot easier, then they can be forcibly included.

**Comment by danielmewes** [comment]
I have to look into it more. We'll probably not have changefeeds in the first revision, and then tackle their integration into GraphQL queries separately.

**Comment by freiksenet** [comment]
Hi!

We at [Reindex](https://www.reindex.io) use GraphQL with RethinkDB, so I'd like to share our experience.

First of all, I agree with @leebyron and others in that GraphQL is not a replacement for ReQL and I don't think RethinkDB implementation should try to replicate all ReQL in GraphQL. However, I think a GraphQL interface to RethinkDB certainly makes sense and Reindex is very similar to it in that it's a generic data storage solution, not a custom backend. 

I think it would make sense to have a generic spec, similar to Relay GraphQL spec, that defines how a *generic* data storage provider should implement basic GraphQL CRUD. This way users using Reindex can switch to RethinkDB GraphQL and to some other provider GraphQL.

What I mean by generic, is that it won't have domain specific GraphQL root fields, but rather have root fields that allow basic operations on types that are defined by GraphQL schema. It will also provide guidelines/proposals on how such data will be stored in the data storage. Spec should be compatible with Relay spec, because Relay is the only (or at least most used) client library for GraphQL. To support Relay some more stuff beyond spec in, eg, mutation payloads are required.

As we already have such generic implementation at Reindex, I humbly propose that we start with it as a basis for such spec.

## What questions should standard answer

Not exhaustive. I'll also provide how Reindex currently answers them.

1. *How to handle schema?*
We store schema in one table in RethinkDB, we generate basic types, connection types, root fields and input objects from it on runtime.
Here is a (very alpha) docs for our type JSON:
https://www.reindex.io/docs/reindex-schema/
2. *How to represent GraphQL types?*
Reindex represents any type with interface `Node` (which is an interface from Relay spec that has unique id) as a RethinkDB table and assumes all other types will be stored inline.
3. *How to represent Relay Connections?*
We define any 1-* relationship between any two `Nodes` a connection, with "parent" side having a Connection field. Non-node types are assumed to be inline, either as lists or objects.
4. *What root fields should be provided?*
We provide CRUD fields (for type `Foo`: `getFoo`, `createFoo`, `updateFoo`, `replaceFoo`, `deleteFoo`)
5. *What is required to support Relay beyond the spec?*
   1. `node` root field that retrieves object of *any* `Node` type by unique id. We basically concat typename and RethinkDB id through a custom scalar type, base64 encode it and use that as a global id.
   2. mutation results (`Payloads`) should have (so that they work with all Relay mutations)
     1. The changed object itself
     2. The changed object Relay `edge`
     3. Field containing `viewer` object
     4. A field for each connection this object can be a part of
  3. "global" connections (so connections to all objects of some type) should be wrapped in an object of `Node` type. We have `Viewer` type and `viewer` root field that returns a singleton objects that has fields containing connections to all types available. It also holds reference to current user.
6. *How user authentication should be handled?*
I think actual authentication is out of scope of the spec, but handling authenticated user is in scope. We pass logged-in user id in GraphQL context to all queries and then it can be used to check permissions. Optionally one can also have some sort of User table and match the logged-in user id with it. We do provide it by default.
7. *How permissions should be handled?*
With the user id passed to context, one can use it to implement permissions. We have Permissions type/table in our system, which stores objects with user id and type id in them, along with CRUD permissions' booleans. Root fields check for permissions and return null if there are none.
We also provide permission propagation through connections, where connections to `User` type (which we have), can grant additional permissions, so you can, for example, model that user can update something that he is an author of. The latter is an experiment, we are not sure how well it will work.

Hope that helps. Looking forward to feedback.

**Comment by tracker1** [comment]
Just one point of input on this.. would suggest to someone building a JS extension for this to support template based query injection for any JS interface...   

    var result = await (r.graphql`
      {
        user(id: ${userId}) {
          name
          age
          friends {
            name
          }
        }
      }
    `)

This would make it more natural to deal with query injection.. yes it's a new feature of JS, and would require babel for most people, it would still be a good idea.  Why this isn't a part of the spec, I don't know.

**Comment by robzhu** [comment]
@coffeemug really excited to see that you guys are fully engaged and taking this seriously. Can't wait to see this feature ship.

**Comment by mquandalle** [comment]
Facebook [published a blog post about GraphQL subscriptions](http://graphql.org/blog/subscriptions-in-graphql-and-relay/) that I believe is relevant in this conversation — especially the part about why they won’t implement a RethinkDB-like live query model.

**Comment by danielmewes** [comment]
@freiksenet Great idea to come up with a common spec. I need to look into the details of this a bit more. Thanks for the writeup!

@mquandalle Very interesting post indeed. Thanks for sharing.

Many of the requirements they describe about the live-query complications happen to be solvable for changefeeds by using changefeeds on table joins (#3997) and on aggregations (#1118).
We have some plans for both of those, so live queries / changefeeds might actually become feasible for a lot of additional queries, including some really complex ones.

**Comment by kofalt** [comment]
I second @mike-marcacci's reservations about directly bundling one or more arbitrary additional query languages. The comment from @coffeemug about containing these in a plugin architecture is reassuring.

As someone very interested in [Falcor](https://netflix.github.io/falcor/) as mentioned upthread, is there room for a more general ticket about supporting graph query languages? (Apologies if one has already been linked). Things like @freiksenet's writeup would be very relevant to that conversation.

The comments in this thread implying a concentration on the support capability first, then language-specific plugins second, makes me a lot less nervous about this.

**Comment by ansarizafar** [comment]
@danielmewes  @coffeemug Any update on the progress so far? Could you please share more information how you are implementing authentication.

**Comment by coffeemug** [comment]
@ansarizafar -- unfortunately no information yet, as we're still in the design stages. I should be able to give some more information in a week or two.

**Comment by coffeemug** [comment]
All -- I've got some good news and some bad news wrt this issue.

The good news is that our preliminary design discussions are complete, and the project to get users direct browser access to RethinkDB is now under way. We expect to ship a complete working feature/product in 6-8 weeks.

The bad news is that this first implementation will not be using GraphQL. We were (and still are) very excited about GraphQL and the possibilities that it opens up for the future of web development, but during the design process we ran into a number of issues, and eventually decided that it would be unwise to offer a GraphQL protocol for the first implementation. Here is a quick summary on how we reached this decision.

I think that the closest mental model for GraphQL is to think of it as a replacement for REST. As we started designing browser access to the database on top of GraphQL we ran into the following problems:

- Relay's GraphQL spec gives relatively little guidance on how to structure the API. For example, there is some guidance on how to structure mutations to make them consumable by Relay, but there is no guidance on which mutations to implement (on this issue GraphQL/Relay is partially less specified than REST). To get a good implementation, we need to figure out how to design basic insert/update/upsert/replace commands in a such a way that React/Relay developers find them useful and convenient, and since GraphQL/Relay is relatively new, it's not immediately clear what a good, convenient mutations API would look like.
- GraphQL has a notion of subscriptions to allow for live updates in the application, but this spec is too narrow for what we want to accomplish. RethinkDB has the capability to do full LiveQuery, which we believe is immensely valuable to application developers, and it's important to us to bake it into the API. If we were to do it in GraphQL, we'd have to extend the spec, fork/hack Relay to operate over WebSockets, and maintain that fork until a full LiveQuery spec/implementation makes it into official GraphQL/Relay. That would be fairly challenging for us to do, and easy to mess up.
- Relay/GraphQL says nothing about authentication and authorization, which is a crucial part of the design for a serverless app product. This isn't specific to GraphQL/Relay -- we'd have to design a good security model/APIs regardless of which protocol we use, but again, since GraphQL/Relay are fairly new and we have little experience with it, it would be very very easy for us to mess up the API.

All of these decisions were made by @leebyron/Facebook's team explicitly and by design, and they all seem very smart to me. For example, if they overspecified the structure of the API, they'd significantly limit applicability of the technology; if they added LiveQuery support prematurely, there would be no practical way for 99% of the users to implement it; and if they added a security model, it wouldn't make very much sense since GraphQL is designed as a general purpose protocol.

The challenge for us, however, is to build a really great API/implementation to allow our users to build engaging, realtime serverless apps, and without having significant experience building GraphQL/Relay applications, this is a sisyphean task (since APIs are extremely easy to mess up).

Here is our plan right now:

- We will build a server-side API/implementation on top of WebSockets/REST -- this will allow us to design a great API because many of us are intimately familiar with these technologies, and the novelty of the protocol will not lead us to make rookie mistakes that will make the API unusable by pragmatic web developers who just want to build their apps.
- We'll also build a JavaScript library that wraps this API on the front-end to let people get started building apps easily (again, this part we understand really well).
- Once the feature is out and the API/security model is solidified, it will be dramatically easier for us to add a native GraphQL adapter. We will almost certainly follow up with a GraphQL implementation of the API after that (once the API is working and we have a lot more information).

Really sorry to disappoint anyone who was hoping to get this feature in the short-term, but it's extremely important to us to design beautiful APIs that let people build engaging applications, and we're super-sensitive to how easy it is to screw that up. So unfortunately a full implementation based on the GraphQL protocol will take a little longer, but the good news is that you can be assured that once it's out, it will be really, *really* good.

**Comment by thelinuxlich** [comment]
Disappoint? This is huge! :D

Is this similar to ArangoDB Foxx?

**Comment by mike-marcacci** [comment]
This sounds like a really well thought-out strategy. I'm really excited and curious about some of these new ideas (especially the security model) but I don't want to take this thread too off-topic with all my new questions :)

I know there are several community-submitted issues/suggestions regarding these great new features: do you know which issues we should follow to stay up-to-date with the official proposals?

**Comment by danielmewes** [comment]
> Is this similar to ArangoDB Foxx?

There are some similarities to Foxx on an abstract level. However we don't aim at moving more logic into the database.
Instead we think that it's important that developers who want to focus on front-end development have to deal with the backend as little as possible (including both the database and potentially an application-specific backend). So one could say that we try to move logic into the frontend code, and make it extremely easy to write reactive applications on that level. Once more control is needed, the backend will be extensible to allow the addition of further logic there, but this will not be necessary to get started with writing a fully functional realtime application.
Of course there are a bunch of security considerations to take into account with this approach. We should have an issue up about that part soon to discuss details.

**Comment by opsb** [comment]
I've been using [rethinkdb-websocket-server](https://github.com/mikemintz/rethinkdb-websocket-server) with [rethinkdb-websocket-client](https://github.com/mikemintz/rethinkdb-websocket-client) which provide a websocket proxy over a rethinkdb connection. If I've understood correctly your plan is to add a baked in version of this, if that's the case then it sounds great!

Curious to see what you're going to do with the security model, [rethinkdb-websocket-server](https://github.com/mikemintz/rethinkdb-websocket-server) is a node server that you can extend with custom logic to approve/deny access to queries. Maybe a rethinkdb plugin would make sense for this?

**Comment by coffeemug** [comment]
> Do you know which issues we should follow to stay up-to-date with the official proposals?

I just created a private repo on GitHub where all discussion and development will take place. We've dedicated three engineers on the RethinkDB team (@deontologician, @dalanmiller, and @Tryneus) to work on the project, with the rest of the RethinkDB engineers advising on all design decisions.

Once we get to a point where we can release the project to the public, it will be released under an open-source license (likely MIT/Apache), and all development will be done in the open with the community just like RethinkDB. The reason we're running a closed development process until we get to v0.1 is that early stages of project development are the most special and vulnerable, and we felt it's important to temporarily limit the number of voices to help ensure success of the project.

Our community team created a co-designer program, and we'll be inviting some of the most active community members to the private github repo to help shape the project while it's at this vulnerable stage. If you'd like to participate in the design process, fill this out -- https://rethinkdb.wufoo.com/forms/rethinkdb-fusion-codesigner-program/ -- and we'll see if we can add you to the repo.

I'm really excited to get this out -- I think it will be amazing. The design process is a ton of fun, but waiting six weeks until we can release it is torture :smile:

**Comment by rattrayalex** [comment]
All of the above sounds fantastic to me. Great leadership!

**Comment by jkarneges** [comment]
@coffeemug , you may recall our discussion about [LiveResource](http://liveresource.org) as a protocol for exactly this problem. Maybe time to resume that. ;)

**Comment by chrisabrams** [comment]
@coffeemug Looking forward to seeing the progress - I will be your #1 beta tester :)

**Comment by 0004c** [comment]
Any fresh news? Btw very interesting to know really, as @opsb said, about rethinkdb-websocket-server and rethinkdb-websocket-client or react-rethinkdb?

**Comment by danielmewes** [comment]
@0004c we're still working heavily on the project @coffeemug mentioned above. It's not quite in a testable state yet, but I think it will be there soon.

The idea for security is similar to @mikemintz 's rethinkdb-websocket-server only in that the browser-accessible layer for RethinkDB is going to run in its own isolated process (comparable to what's described in https://github.com/rethinkdb/rethinkdb/issues/4785 as you mention).
The security model itself is quite different though. Our server won't actually expose ReQL as its query language directly to the client. Instead it will provide a custom-built API that makes the common functionality easily accessible, while still providing simple extensibility for advanced use cases. We should have more to share in the next weeks.

**Comment by natew** [comment]
@danielmewes sounds interesting and would love to talk with you guys more about it. I am curious around the write and subscribe areas especially. With recent progress on subscriptions in relay as well, and your general approach implementing the various features (read, write, subscribe).

**Comment by danielmewes** [comment]
@natew Would you mind pinging me on our Slack?
http://slack.rethinkdb.com/
I'm danielmewes over there.

**Comment by ansarizafar** [comment]
@coffeemug  I guess eight weeks period is over, Any new year update?

**Comment by coffeemug** [comment]
We're still on this, making good progress (as usual, humans suck at estimates -- this will take longer than expected). It's starting to be amazing, I'm really excited to get this out!

**Comment by niieani** [comment]
Sounds awesome! Happy new year guys.

**Comment by rnenjoy** [comment]
Cant wait to see the progress of this!

**Comment by nbostrom** [comment]
Related discussion in meteor land: https://forums.meteor.com/t/conversation-in-place-of-livequery-how-should-meteor-build-webserver-tier-subscriptions-i-e-for-graphql-relay/16844/2

**Comment by ansarizafar** [comment]
@coffeemug Is it possible to join beta group?

**Comment by mglukhovsky** [comment]
@ansarizafar: you can email me at mike@rethinkdb.com, we'll set it up.

**Comment by ansarizafar** [comment]
@mglukhovsky I had emailed you day before yesterday. Please let me How can I get access. Thanks.

**Comment by coffeemug** [comment]
@ansarizafar -- just invited you. Sorry for the delay, there is more activity around the project than we expected, so this slipped through the cracks. /cc @mglukhovsky

**Comment by ansarizafar** [comment]
@coffeemug @mglukhovsky Thanks, I am really excited about this project.

**Comment by alexgorbatchev** [comment]
@coffeemug could you please post an update? :coffee:

**Comment by luiscarli** [comment]
This project may be relevant for this discussion: https://github.com/mikemintz/rethinkdb-websocket-server
Together with: https://github.com/mikemintz/rethinkdb-websocket-client

It allows for doing the queries directly on the front-end side, and incrementally adding query validation as the project progress.

**Comment by Dev1an** [comment]
Meteor development group (MDG) is doing some interesting work concerning this topic: https://github.com/apollostack/apollo
They make a javascript stack for realtime data subscriptions using GraphQL. Would it be a possibility to work together with them and implement a [Data driver](https://github.com/apollostack/apollo/blob/master/design/high-level-reactivity.md#data-drivers) so that we can do realtime graphQL queries to a rethinkDB via Apollo?

**Comment by mattkrick** [comment]
@mglukhovsky can i get an invite?

**Comment by mglukhovsky** [comment]
@mattkrick -- sent you an email last night with details (and you should be invited to the project.)

**Comment by niieani** [comment]
@mglukhovsky Would love an invite too. I've been developing my own optimistically-updating client-side querying lib with @mikemintz's `rethinkdb-websocket-client` and it's been great so far - I'd love to see what the officially supported, like-minded GraphQL with RethinkDB has to offer.

**Comment by mattkrick** [comment]
@mglukhovsky still getting a 404 on the horizon repo

**Comment by mglukhovsky** [comment]
@mattkrick: check your email for a GitHub invite. You can view and accept invitations here: https://github.com/rethinkdb

**Comment by mglukhovsky** [comment]
@niieani: could you send me an email at mike@rethinkdb.com? Happy to get you set up.

**Comment by luiscarli** [comment]
@mglukhovsky can i get an invite also? Thanks!

**Comment by usergit** [comment]
may I get an invite? also it would be awesome if you guys post a guideline or a sample app that shows how to handle authorizations. thanks

**Comment by niallobrien** [comment]
@mglukhovsky @coffeemug Can I get an invite please? Thanks in advance.

**Comment by Dev1an** [comment]
@mglukhovsky  @coffeemug Can I get an invite?

**Comment by nbostrom** [comment]
I would really like one too! This would be perfect for my next project!

**Comment by mglukhovsky** [comment]
@luiscarli, @niallobrien, @Dev1an, @nbostrom, sent you an email + invite.

@usergit, could you send me an email at mike@rethinkdb.com?

**Comment by fredp46** [comment]
@mglukhovsky @coffeemug I'd love an invite too.

**Comment by apucacao** [comment]
@mglukhovsky I would really like an invite too. Thanks!

**Comment by jbroadway** [comment]
@mglukhovsky @coffeemug I would love one too! (lurking on this thread for some time in anticipation :)

**Comment by tonyxiao** [comment]
I'm confused, what is the invite to? slack team or rethinkdb's beta support for graphql?

**Comment by alexgorbatchev** [comment]
Please invite me as well, thank you!

**Comment by coffeemug** [comment]
All -- quick update.

The project is moving along and we settled on the name `horizon`. Our target ship date for Horizon beta is April 15th. We're very excited about this project because it will make building advanced, modern apps dramatically easier.

The first version will not have GraphQL support, but Horizon was designed with GraphQL in mind, so we should be able to follow up with a GraphQL adapter pretty quickly (possibly in the next version after launch, depending on the feature list).

We will likely open up the GitHub project before the launch date -- we're currently working on a security model/implementation, and once that lands we'll open source everything and make it generally available.

Thanks everyone for your support -- we're super excited about shipping this! (the first 90% is done, we're now working on the last 90% :smile: )

**Comment by coffeemug** [comment]
@tonyxiao -- the invite is to the Horizon project (currently and temporarily a private GitHub project) and the community on Slack. 

/cc @mglukhovsky -- could you please invite the folks above?

**Comment by thelinuxlich** [comment]
@coffeemug Is this project compatible with any databinding framework? I hope to use it with Vue.js

**Comment by coffeemug** [comment]
Horizon is agnostic to databinding frameworks -- you should be able to use a framework of choice fairly easily.

**Comment by mglukhovsky** [comment]
For those of you looking to get a Horizon invite, you can also find me (@mglukhov) on the #horizon channel on our [community Slack](slack.rethinkdb.com), along with the rest of the team.

**Comment by ktersius** [comment]
@mglukhovsky I'm working on a pet project of mine and I'd really be interested to see what you guys have planned with Horizon. It might affect some of my design decisions... Can I also get an invite please?

**Comment by mglukhovsky** [comment]
@ktersius: could you ping me on our Slack channel, or send me an email? (mike@rethinkdb.com)

**Comment by threepointone** [comment]
Would you please add me too? Thank you!

**Comment by cymen** [comment]
I'd love an invite too. Thanks!

**Comment by chentsulin** [comment]
May I also get an inivite? Thanks!

**Comment by webigor44** [comment]
And me, thanks!

**Comment by threepointone** [comment]
I apologize, shouldn't have asked here. Followed up on slack.

**Comment by byw** [comment]
Could you add me as well? Thanks!

**Comment by TofPlay** [comment]
Me too. Thanks 😊

**Comment by brownish** [comment]
I would love an invite as well. Sounds great!

**Comment by believer** [comment]
I would love an invite! Really excited about this! :smile:

**Comment by rnenjoy** [comment]
Me too :)

**Comment by knownasilya** [comment]
@coffeemug if you can spare an invite :+1:

**Comment by mquandalle** [comment]
I have subscribed to this thread because I’d like to follow RethinkDB+GraphQL developments, but I’m getting a bit bored to be notified everytime someone requests an invite. Could you consider another process for invite requests (like a google form, or writing your name on a etherpad document, or something else)?

**Comment by knownasilya** [comment]
Or whoever upvotes your comment?

**Comment by stream7** [comment]
Or people could join the slack team first http://slack.rethinkdb.com/ and ask for an invite there?

**Comment by coffeemug** [comment]
I'm thinking through another process with @mglukhovsky right now. We tried Slack and a web form, but it's getting lost in the comments. We should figure out how to handle this better in a day or two. Sorry everyone for getting spammed with invite requests!

**Comment by davidbilly** [comment]
@coffeemug Hi, could you invite me into Horizon project. I been asked on slark group, but nobody reply me :( Thanks!

**Comment by mglukhovsky** [comment]
Hey guys, we had an unexpected level of interest in this project. We're starting to open the developer preview for Horizon (though it's still early.)

You can sign up here: https://horizon.io

I'd like to keep the volume of traffic on this conversation minimal for existing participants, so I'm going to lock this conversation for the next few weeks to collaborators. I'll unlock it as Horizon nears general availability. We'll post updates as they come -- thanks for your patience!
